api/types.go

```diff
@@ -1106,17 +1106,36 @@ func (pin Pin) ProtoMarshal() ([]byte, error) {
 		timestampProto = uint64(pin.Timestamp.Unix())
 	}
 
+	// Our metadata needs to always be seralized in exactly the same way,
+	// and that is why we use an array sorted by key and deprecated using
+	// a protobuf map.
+	var sortedMetadata []*pb.Metadata
+	var metaKeys []string
+	for k := range pin.Metadata {
+		metaKeys = append(metaKeys, k)
+	}
+	sort.Strings(metaKeys)
+
+	for _, k := range metaKeys {
+		metadata := &pb.Metadata{
+			Key:   k,
+			Value: pin.Metadata[k],
+		}
+		sortedMetadata = append(sortedMetadata, metadata)
+	}
+
 	opts := &pb.PinOptions{
 		ReplicationFactorMin: int32(pin.ReplicationFactorMin),
 		ReplicationFactorMax: int32(pin.ReplicationFactorMax),
 		Name:                 pin.Name,
 		ShardSize:            pin.ShardSize,
-		Metadata:             pin.Metadata,
-		PinUpdate:            pin.PinUpdate.Bytes(),
-		ExpireAt:             expireAtProto,
+		// Metadata:             pin.Metadata,
```

Comment:
Peers on older versions will not be able to see metadata  for pins submitted on newer versions.